File: src/neetcode/Java/_46Permutations.java
package neetcode.Java;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class _46Permutations {
    List<List<Integer>> list = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        ArrayList li = new ArrayList<>();
        for (int i : nums) li.add(i);
        Helper(li, new ArrayList<>());
        return list;
    }

    public void Helper(ArrayList<Integer> nums, ArrayList<Integer> l) {
        if (nums.size() == 0) {
            list.add(new ArrayList<>(l));
            return;
        }
        for (int n = 0; n < nums.size(); n++) {
            int val = nums.get(n);
            l.add(val);
            nums.remove(n);
            Helper(nums, l);
            nums.add(n, val);
            l.remove(l.size() - 1);
        }
    }

    public static void main(String[] args) {
        _46Permutations obj = new _46Permutations();
        int[] nums = {1, 2, 3};
        System.out.println(obj.permute(nums));
    }
}

File: src/neetcode/kotlin/_46Permutations.kt
package neetcode.kotlin

class _46Permutations {
    var list: MutableList<List<Int>> = ArrayList()

    fun permute(nums: IntArray): List<List<Int>> {
        val li: ArrayList<*> = ArrayList<Any>()
        for (i in nums) li.add(i as Nothing)
        Helper(li, ArrayList())
        return list
    }

    fun Helper(nums: ArrayList<*>, l: ArrayList<Int>) {
        if (nums.size == 0) {
            list.add(ArrayList(l))
            return
        }
        for (n in nums.indices) {
            val `val` = nums[n]
            l.add(`val` as Int)
            nums.removeAt(n)
            Helper(nums, l)
            nums.add(n, `val` as Nothing)
            l.removeAt(l.size - 1)
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _46Permutations()
            val nums = intArrayOf(1, 2, 3)
            println(obj.permute(nums))
        }
    }
}
